File: 백준/Gold/2470. 두 용액/두 용액.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
const n = +input[0];
const arr = input[1].split(' ').map(Number);

arr.sort((a, b) => a - b);
let left = 0;
let right = n - 1;
let minSum = Infinity;
let lIdx = left;
let rIdx = right;

while (left < right) {
  const sum = arr[left] + arr[right];

  if (Math.abs(sum) < Math.abs(minSum)) {
    minSum = sum;
    lIdx = left;
    rIdx = right;
  }

  if (sum < 0) left++;
  else if (sum > 0) right--;
  else break;
}

console.log(arr[lIdx], arr[rIdx]);

File: 백준/Gold/2470. 두 용액/두 용액.py
import sys
from json.encoder import INFINITY

input = sys.stdin.readline
n = int(input())
arr = list(map(int, input().split()))
arr.sort()
res = [0, 0]
min_val = INFINITY
l, r = 0, n - 1

while l < r:
    a = arr[l] + arr[r]
    if abs(a) < min_val:
        min_val = abs(a)
        res = [arr[l], arr[r]]
    if a > 0:
        r -= 1
    elif a < 0:
        l += 1
    else:  # 합이 정확히 0일땐 종료
        print(arr[l], arr[r])
        exit()

print(res[0], res[1])
